### ISSUE TITLE
Hierarchical Transitions

### ISSUE THREAD

**Issue by stuartmscott** [AletheiaWareLLC/Wink#45]
State Transitions should move through the hierarchy, so that leaving the old state exits all the parent states going up the tree until the common ancestor (if any), and then enter all the parent states going down the tree and into the new state.

Ie. given a State Machine:

```
         parent
         /    \
        /      \
      leaf1    child1
               /   \
              /     \
            leaf2  leaf3
```

- Transitioning from `leaf2` to `leaf3` results in the execution of `leaf2.OnExit` and `leaf3.OnEntry`
- Transitioning from `leaf1` to `leaf3` *should* result in the execution of `leaf1.OnExit`, `child1.OnEntry`, and `leaf3.OnEntry`
- Transitioning from `leaf3` to `leaf1` *should* result in the execution of `leaf3.OnExit`, `child1.OnExit`, and `leaf1.OnEntry`

**Comment by winksaville** [comment]
When implemented, my suggestion is to add onEntryCounter, onExitCounter and onReceiverCounter for each state, probably always for debug builds and optional for release builds, and then use those counters to validate the when in what order the methods are executed.

**Comment by winksaville** [comment]
> When implemented, my suggestion is to add onEntryCounter, onExitCounter and onReceiverCounter for each state, probably always for debug builds and optional for release builds, and then use those counters to validate the when in what order the methods are executed.

Which is issue #33

**Comment by stuartmscott** [comment]
Yup, I've already used this approach in [testing](https://github.com/AletheiaWareLLC/Wink/blob/main/test/src/machine.cpp#L68), but agree having them integrated into `Machine` would be best.

I like the idea of only having them in debug builds, but I haven't setup build variants yet so the counters will probably be always-on and we can revisit in future.

Also: https://github.com/AletheiaWareLLC/Wink/issues/46
